### DISCUSSION TITLE
System tiddlers content hash, support updateas, patches

### DISCUSSION THREAD

**Discussion by AnthonyMuscio** [TiddlyWiki/TiddlyWiki5#5954] [Ideas]
Folks,

I was wondering if we could build a **System tiddler content hash** for tiddlers in the core?. This would be stored in a hidden field on each system tiddler, or regular ones by design. I believe it serves a few identified purposes below but such a facility could be used in other applications if made available as widgets.

Basically as a new release is issued an algoritium runs against each system tiddler which is analysed and generates a hash, ideally using a method to keep the hash size as small as possible (eg use a base 60 encoding) whilst giving those interested the ability to **test if the system tiddler has changed.** Yes this is in addition to shadows vs tiddlers. Basically a good hash will allow you to detect if even one character in the tiddler changes. Ideally this would apply to the whole tiddler but text would be enough.

Now if the hashing algorithm is made available to us, we can test a given system tiddler for changes, if the stored (published) hash does not match the computed hash. This could be used **for the upgrade utility to quarantine edited system tiddlers** that have being modified, and Plugins forced to modify a system tiddler, could **detect if an upgrade has changed an underlying system tiddler** it wants to modify. Ideally this would **trigger a diff view** one could choose **to update or maintain** the changes selectively.

Most of all such a facility could give rise to the **ability to distribute core hacks or patches** which can detect if and when a **plugin or upgrade has changed a system or shadow tiddler.** This will be of great benefit to designers and developers who want to test and share possible core changes with a patch.

Plugin authors could detect changes to their own tiddlers and include handling as desired. Some beautiful Agile and component based development solutions could be built on this alone, not to mention the ability to use this to manage data plugins and as a de facto packable delete, retain the hash but delete the content.

Your thoughts?
Tony

**Comment by AnthonyMuscio** [comment]
Post script:

Sure a plugin designed to do this could be made but it would change all shadows, that's why it should be slipped into the core or a special edition.

Importing tiddlers whose has hash not changed could be discarded and it made clearer to the importer which tiddlers actually changed on import/overwrite. 

Perhaps modified date needs to be left out of the hash.

Tony

**Comment by pmario** [comment]
The simple check, if a system tiddler is overwritten is: `[is[tiddler]is[shadow]sort[title]]`. This is part of the $:/ControlPanel : Info : Basics tab. ... BUT ... This info doesn't really show if the tiddler content has been changed. It could have been opened and saved without changing something.

------

>I was wondering if we could build a System tiddler content hash for tiddlers in the core?. This would be stored in a hidden field on each system tiddler, or regular ones by design

That doesn't work. ... If the hash is part of the system tiddler I can create a plugin, that overwrites a core tiddler and add the hash of the new tiddler to that field. ... Since the hash-field can _not_ be part of the hash you have no chance to detect, that I did create a new system tiddler, that looks like core, because the hash is valid. ... We can only trust a hash, if it comes from a "trusted source" outside the wiki.

The only way that works is, that the wiki build process will create the hashes and publish them separately and versioned. So if a user or a plugin author wants to check the code has been modified, they would calculate the hash of the wiki-tiddler and compare it with the hash of the original-tiddler published at GitHub. 

-----

If we use a [sha256](https://en.wikipedia.org/wiki/SHA-2) hash, which should be secure for now, it has 32 characters. Changing the encoding will make validation more complicated.

**Reply by AnthonyMuscio** [reply]
Mario,

The purpose of my proposal is not to secure tiddlers against edit, I accept when a shadow becomes a tiddler, the hash can be updated etc... The intention is really just to test if an actual change has occurred in the tiddler through a simple edit. The hash would be used more like a CRC - cyclic redundancy check.

If for example the update process is about to replace a core tiddler, if it discovers it has a tiddler (overwritten shadow) and it checks a newly minted hash with the hash field value (put there in the last release) it can determine if a substantive edit has occurred.  The tiddler can be moved into a quarantine JSON of tiddlers, and after the update it can report all tiddlers so quarantined and replaced for manual review (with a dif and compare tool). Since the new core tiddlers that have changed now have a different hash, a plugin could determine that the core tiddler has changed underneath it - eg may need a repatch and could prompt the user to repatch/edit the now different core tiddler.

**Reply by AnthonyMuscio** [reply]
Can I ask what character set does sha256 use? 0-9-A-Z?

**Comment by pmario** [comment]
>... Yes this is in addition to shadows vs tiddlers. Basically a good hash will allow you to detect if even one character in the tiddler changes. Ideally this would apply to the whole tiddler but text would be enough.

IMO "text" would not be enough. TW allows us to use a `type-field: application/javascript` ... which makes a tiddler an executable. So without changing the text we can completely change the behaviour of a tiddler. So the hash would need to include everything.

**Reply by AnthonyMuscio** [reply]
Agreed, but I am not trying to protect tiddlers, just detect when they are different, we can decide which if any fields are included in the hash, thus if edited can be detected, eg tags or as you suggest the type field.

**Comment by pmario** [comment]
For plugins, it would be enough to have 1 hash for the whole plugin including _all_ tiddlers.

**Reply by AnthonyMuscio** [reply]
I don't agree, what if I only change one tiddler in a plugin. In fact it may be a config tiddler. I would like a way to detect which tiddler was edited.

**Reply by pmario** [reply]
If you change a config tiddler, that comes with a plugin and you change the tiddler, doesn't change the plugin itself. Tiddlers in a plugins are "extracted" at startup. So you can easily compare tiddlers that have been changed. see: 

 - `[is[tiddler]is[shadow]sort[title]]` 
 - You can test that at tiddlywiki.com
 - Open https://tiddlywiki.com/#%24%3A%2FDefaultTiddlers

see: 
![grafik](https://user-images.githubusercontent.com/374655/129479474-c0d3a3f9-ada1-46d6-bdce-0034d322d8fd.png)

**Reply by pmario** [reply]
If you can trust the hash of the plugin, you can safely compare the tiddler content using default tools

**Comment by pmario** [comment]
>... We can only trust a hash, if it comes from a "trusted source" outside the wiki.

This gives us a new problem. .. How can we verify, that nobody did mess with the hashes? ... Compress the hashes into 1 file and sign it with PGP. 

Similar to:  https://wiki.debian.org/Creating%20signed%20GitHub%20releases

**Reply by AnthonyMuscio** [reply]
Sorry, what I mean here, and may have misled you. Is lets generate the hashes when publishing a new version of the wiki eg; 5.2.1 these changes will remain the same indefinitely until a future version makes a change and the hash changes.  My trust here not that it has not being maliciously tampered with but the hash represents the published version.

**Reply by AnthonyMuscio** [reply]
A record of hashes at version publishing could be held in a separate plugin and lookups done but this is more complex, although the core need not be changed.

**Comment by AnthonyMuscio** [comment]
Sorry, I am still working through the whole story, but if you can see my reasons, perhaps we can find a solution.

So my idea is that perhaps the vast majority of core tiddlers do not change between versions, if a plugin applied a patch to a system tiddler with a particular hash then if that hash does not change then the core tiddler has not changed and the same patch can be retained or reapplied. If the core hash has being changed then all bets are off and the patch cant be reapplied without intervention.

In some ways this would allow every tiddler in the core to have its own change flag, even a hash for each version giving finer grained control and the ability to write patches to a specific "version" or "hash" of a core tiddler.

**Reply by pmario** [reply]
I can see, what you want to achieve. It works, but it doesn't scale. ... It's called [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch) ... We try to avoid it as good as possible. ... Using hashes still makes it a monkey patch.

Let me explain. 

 - Plugin A checks a core function and sees it is able to modify it.
 - It modifies eg: `CheckboxWidget.prototype.render()` function
 - Plugin B checks the core and sees it can modify it. 
 - It modifies `CheckboxWidget.prototype.getValue()`

Since this are different elements of the same widgets it may look like there is no problem. BUT `render()` uses `getValue()` so we have 2 plugins now, that modified a core function in unpredictable ways. ... This mechanism is an active "time bomb". ...

Now let's assume both of them modify the same `render()` function. So the last one who modified it wins. .. In sum the functionality is broken, for at least 1 plugin

The same flow is valid for non-js-tiddlers. We have the same problems here. 

That's why since 5.2.20 the core allows us to [subclass](https://tiddlywiki.com/dev/#WidgetSubclassingMechanism) widgets. There is still a chance, that a future core version of the widget will break the plugin widget ... but ... it will **only** break the plugin widget, because it has a different name. 

Your approach would break the core. 

----------

The core also provides [hooks](https://tiddlywiki.com/dev/#HookMechanism) to allow plugin authors to modify tiddlers at different stages, when they are changed and written to the store. The difference to your approach is, that this mechanism works for every plugin author.

TW-hooks stack. So several plugins can use the same hook and modify the content in a well defined way.

**Reply by AnthonyMuscio** [reply]
Mario, I don't write javascript (yet). I see I could have stated that. I am talking about patching wikitext tiddlers ONLY.

But I will give this some more thought. I hear what you are saying but I am not sure I have expressed myself in the right language.

An example may be adding "actions=<<action-variable>>" to core buttons, or wrapping the display of the tiddler body with an additional variable to use inside all tiddlers, transcluding some buttons in front of a list or tiddler titles etc...

Eg; edit $:/core/macros/timeline to include {{||$:/core/ui/Buttons/edit}}
as follows 
`{{||$:/core/ui/Buttons/edit}} <$link to={{!!title}}><<timeline-title>></$link>`

The above adds an edit button in front of all items in the recent tab. Now my plugin could just replace the $:/core/macros/timeline tiddler but if the timeline macro is updated in the next release, it will be either hiding under in the shadow or replaced and the change lost. I want to learn when modified tiddlers are hiding core updates or let a plugin know when the core tiddler has changed and automate the possible reapplication/repatch of that change.

**Reply by pmario** [reply]
See: #5947 and #5948 ... IMO that's the way to go for "core tiddlers", to make them more "hackable". 

As I wrote, a plugin can _not_ patch a core tiddler without changing it. .. The last plugin that changes a core tiddler wins. ... That's a mechanism, that doesn't scale. ... 

core-UI elements can only be made more flexible, if they provide macros, that can be overwritten by plugin authors. ... But still the "last one wins" will always be there. 

~~We don't have a mechanism atm that would allow us to inherit a UI element and apply changes to it, where the next plugin could do the same thing ... and so on.~~

While writing the last sentence I did think about the `\import` pragma. ... It may provide us with this mechanism... We just don't make it usable in the core atm.

**Reply by AnthonyMuscio** [reply]
Thanks for staying with me on this. I will have another go at explaining. It is from my super-user perspective, not that of a developer or core developer.

To clarify in this comment the shadow tiddler is the "immutable" one, the edited copy is a "standard tiddler" of the same title. 

Currently a user, or recently installed plugin can effectively edit a shadow tiddler, a core or plugin shadow. It does this by creating a "standard tiddler" with the same name, and this is the last that one wins, ie the "standard tiddler" will be used rather than the shadow. 

A "standard tiddler" acts as a "shadow tiddler overlay", it may be created with no change to the shadow below it, with an edit and save, ie without changing anything. In other words "it looks like its being changed but it has not", however now the "source of truth" is this tiddler. Of course the "standard tiddler" can also have modifications to change how tiddlywiki works/looks, it may be as simple as an additional tag, this then becomes the source of truth.

Now if we upgrade a tiddlywiki, some of the core tiddlers are changed (underneath) but the "standard tiddler" of the same name is not. the result is the source of truth continues to be the overridden "standard tiddler" version. If the shadow tiddler introduced new features, these will be unavailable and in fact using the old version in the "standard tiddler" could cause problems.

So my idea was, when building shadow tiddlers we compute a hash of the tiddler and store it in the tiddler itself (a hidden field). Now if we edit the shadow tiddler and create a "standard tiddler" by the same name it will inherit the published hash. If we edit the tiddler and make a change, if we re-computed the hash again (but don't save it in the hash field) we would discover it does not match the hash in the "standard tiddler" or the "Shadow Tiddler", this allows us to confirm a real change was made. 

If we upgraded the wiki and the underlying "shadow tiddler" was updated in the new release, it would have a new hash calculated and stored in its hash field. The first thing that would be clear is the "Shadow tiddler hash" and the "Standard tiddler hash" of the same name will be different. This tells us the core has being upgraded under the current "standard tiddler" of the same name. 

As part of the upgrade process when we see a "standard tiddler" exists we can check the hash, if it does not vary we can perhaps list it but, otherwise ignore it, any edits will remain in place. If the hash is different it means the "standard tiddler" is based on an older "shadow tiddler" in which case it will in effect suppress the new shadow tiddler (not just that it has an edit). 

In this case we could first calculate the hash for the "Standard tiddler" again and see if it has really changed, if not discard it, if it has actually changed "move" the "standard tiddler" into a JSON tiddler (lets call this Quarantined) and the new shadow tiddler will become the source of truth.

So at the end of an upgrade, especially if something is not working, the user could review all the "quarantined" tiddlers, and even use a difference check. They will see both the impact of the upgrade and the previous changes in the "standard tiddler" and can choose to apply them again if desired.

In the mean time a plugin could be designed to replace or override only shadow tiddlers with a specific hash, because if the hash changes then all bets are off. Ie if it is trying to "patch" a "shadow tiddler". The plugin itself could report "this patch was for a different version" of this tiddler name.

I this proposed solution all that is needed is storing the hash when a new version of tiddlywiki is generated and if the tiddler was changed. Providing a macro or widget to test or get the hash made available, and can be used by designers and plugin writers if needed. This once a generation has passed the upgrade and import mechanisms can be enhanced with this information available.

Does this make a little more sense?

**Reply by pmario** [reply]
OK, I see, what you want. .. There is an argument, that you loose the "old system-hash" once the core is updated. But this can all be done without the need to change the core. 

- We do have `th-saving-tiddler`, which allows us to insert the field into an edited shadow tiddler, when it will be saved. 
- We do have `$:/config/EditTemplateFields/Visibility/`, which allows a plugin to hide certain fields in the edit-template

With 5.2.0 we will have all the th-import hooks in place to validate the new info. But implementing this into the "Import UI" will be really painful.

**Reply by AnthonyMuscio** [reply]
Mario,

Your last response makes me wonder if you see what I want because how will you do it without the core?

I see that the tiddler save mechanism and visibility would be part of the solution, but first we need a hash algorithm we can run against any tiddler such that a single character change will result in a different hash. Now we can apply these to every tiddler but they will change from shadows to tiddlers, alternatively we could store them separately in a data tiddler with the tiddler title as key, however I would think the published core hashes need to be in place before any plugin or patch can make use of it. Each empty that is released we could generate a hash database?

Perhaps we can stop discussion ping pong and have a one on one discussion?. Can you find a time that may work with my UTC +10 and I am sure we can simplify the discussion?

Regards
Tony

**Reply by pmario** [reply]
>I see that the tiddler save mechanism and visibility would be part of the solution, but first we need a hash algorithm we can run against any tiddler such that a single character change will result in a different hash.

Try this at tiddlywiki.com in the F12 browser console. 

```
window.sjcl.hash.sha256.hash("asdfsaaf").map((i) =>{return (i < 0 ? 0xFFFFFFFF + i + 1 : i).toString(16).toUpperCase();}).join("-")
```

**Reply by AnthonyMuscio** [reply]
If this is hashing asdfsaaf into "4CF02349-315A72CE-5CB96C71-B4757A46-43BBC7AF-ABF85604-3CE3611C-EEB37AB5" we may as well retain a copy of the content, rather than a hash, or am I missing something?

**Reply by pmario** [reply]
I did send you a PM with my Skype handle.

**Reply by AnthonyMuscio** [reply]
Mario, I will try and contact you at an appropriate hour in the next few days.
